File: src/main/java/hardcoded/aoc2021/Day5.java
package hardcoded.aoc2021;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import hardcoded.util.Utils;

public class Day5 {
	static class Line {
		final int x1, y1, x2, y2;
		final int minX, minY, maxX, maxY;
		final boolean isAxisBounded;
		
		public Line(String value) {
			String[] parts = value.split("( -> )|,");
			this.x1 = Integer.parseInt(parts[0]);
			this.y1 = Integer.parseInt(parts[1]);
			this.x2 = Integer.parseInt(parts[2]);
			this.y2 = Integer.parseInt(parts[3]);
			this.minX = Math.min(x1, x2);
			this.minY = Math.min(y1, y2);
			this.maxX = Math.max(x1, x2);
			this.maxY = Math.max(y1, y2);
			this.isAxisBounded = (x1 == x2) || (y1 == y2);
		}
		
		void draw(Map<Long, Long> map) {
			if(y1 == y2) {
				for(int x = minX; x <= maxX; x++) {
					long point = getCoordinate(x, minY);
					map.put(point, map.getOrDefault(point, 0L) + 1L);
				}
			} else if(x1 == x2) {
				for(int y = minY; y <= maxY; y++) {
					long point = getCoordinate(minX, y);
					map.put(point, map.getOrDefault(point, 0L) + 1L);
				}
			} else {
				int xd = x1 < x2 ? 1:-1;
				int yd = y1 < y2 ? 1:-1;
				int dd = Math.max(maxX - minX, maxY - minY);
				for(int i = 0; i <= dd; i++) {
					long point = getCoordinate(x1 + xd * i, y1 + yd * i);
					map.put(point, map.getOrDefault(point, 0L) + 1L);
				}
			}
		}
	}
	
	private static long getCoordinate(int x, int y) {
		return ((long)(x) & 0xffffffffL) | (((long)y) << 32L);
	}
	
	public static void main(String[] args) throws Exception {
		List<Line> lines = Utils.readAllLines(2021, "day5").stream().map(Line::new).toList();
		
		Utils.printf("Day 5\n");
		Utils.printf("PartOne: %d\n", partOne(lines));
		Utils.printf("PartTwo: %d\n", partTwo(lines));
	}
	
	public static int partOne(List<Line> lines) throws Exception {
		final Map<Long, Long> system = new HashMap<>();
		
		for(Line line : lines) {
			if(line.isAxisBounded) {
				line.draw(system);
			}
		}
		
		int result = system.values().stream().mapToInt(i -> (i > 1) ? 1:0).sum();
		return result;
	}
	
	public static int partTwo(List<Line> lines) throws Exception {
		final Map<Long, Long> system = new HashMap<>();
		
		for(Line line : lines) {
			line.draw(system);
		}
		
		int result = system.values().stream().mapToInt(i -> (i > 1) ? 1:0).sum();
		return result;
	}
}

File: src/main/python/Day5.py
import Utils as Utils
import re

class Line:
    def __init__(self, value):
        parts = re.split(' -> |,', value)
        self.x1 = int(parts[0])
        self.y1 = int(parts[1])
        self.x2 = int(parts[2])
        self.y2 = int(parts[3])
        self.minX = min(self.x1, self.x2)
        self.minY = min(self.y1, self.y2)
        self.maxX = max(self.x1, self.x2)
        self.maxY = max(self.y1, self.y2)
        self.isAxisBounded = (self.x1 == self.x2) or (self.y1 == self.y2)
        
def increment(system, point):
    system[point] = system.get(point, 0) + 1

def draw(line, system):
    if line.y1 == line.y2:
        for x in range(line.minX, line.maxX + 1):
            increment(system, (x, line.minY))
    elif line.x1 == line.x2:
        for y in range(line.minY, line.maxY + 1):
            increment(system, (line.minX, y))
    else:
        xd = 1 if line.x1 < line.x2 else -1
        yd = 1 if line.y1 < line.y2 else -1
        dd = max(line.maxX - line.minX, line.maxY - line.minY)
        for i in range(0, dd + 1):
            increment(system, (line.x1 + xd * i, line.y1 + yd * i))

@Utils.time_it_and_evaluate
def partOne(lines):
    system = {}
    for line in lines:
        if line.isAxisBounded:
            draw(line, system)
    
    return sum(1 if v > 1 else 0 for k, v in system.items())

@Utils.time_it_and_evaluate
def partTwo(lines):
    system = {}
    for line in lines:
        draw(line, system)
    
    return sum(1 if v > 1 else 0 for k, v in system.items())

if __name__ == '__main__':
    lines = [ Line(x) for x in Utils.readAllLines(2021, 'day5') ]
    
    print('Day 5')
    partOne(lines)
    partTwo(lines)